### PR TITLE
Fix RegexBoyerMoore.IsMatch case-insensitive (regression from 2.1 to 3.0)

### LINKED ISSUE #32534: Regex "^aA$" does not match "aA" for danish culture using IgnoreCase in 3.1
In 3.1 (tested with 3.1.2 on Windows) there seem to be a regression in regular expression matching for danish culture when using IgnoreCase. 

For example, for this program:

```csharp
public static void Main(string[] args)
{
    Match("^aA$", "aA", "da-DK", RegexOptions.IgnoreCase);
    Match("^aa$", "aA", "da-DK", RegexOptions.IgnoreCase);
    Match("^aA$", "aA", "da-DK", RegexOptions.IgnoreCase | RegexOptions.Compiled);
    Match("^aa$", "aA", "da-DK", RegexOptions.IgnoreCase | RegexOptions.Compiled);
    Match("^aA$", "aA", "da-DK", RegexOptions.None);
    Match("^aa$", "aA", "da-DK", RegexOptions.None);
}

public static void Match(string pattern, string text, string culture, RegexOptions options)
{
    CultureInfo.CurrentCulture = new CultureInfo(culture);

    var regex = new Regex(pattern, options);

    Console.WriteLine($"{pattern}/{text}/{culture}/{options}: {regex.IsMatch(text)}");
}
````

When targeting netcoreapp2.1 (or net461) this prints:

```
^aA$/aA/da-DK/IgnoreCase: True
^aa$/aA/da-DK/IgnoreCase: True
^aA$/aA/da-DK/IgnoreCase, Compiled: True
^aa$/aA/da-DK/IgnoreCase, Compiled: True
^aA$/aA/da-DK/None: True
^aa$/aA/da-DK/None: False
```

When targeting netcoreapp3.1 this prints:

```
^aA$/aA/da-DK/IgnoreCase: False
^aa$/aA/da-DK/IgnoreCase: False
^aA$/aA/da-DK/IgnoreCase, Compiled: True
^aa$/aA/da-DK/IgnoreCase, Compiled: True
^aA$/aA/da-DK/None: True
^aa$/aA/da-DK/None: False
```

This seem to be caused by the changes in RegexBoyerMoore in dotnet/corefx#30632, as `string.Compare("aa", 0, "aA", 0, 2, true, new CultureInfo("da-DK"))` returns 1.

### PR DESCRIPTION
**Pull request by stephentoub**
As part of a larger change, a change was made for .NET Core 3.0 that switched a character-by-character comparison loop to instead use string.Compare.  This represents a regression, however, when a culture-aware character-by-character comparison yields different results from a multiple-character comparison.

Fixes https://github.com/dotnet/runtime/issues/32534
cc: @ViktorHofer, @tarekgh, @eerhardt, @pgovind

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexBoyerMoore.cs
Comment (tarekgh) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexBoyerMoore.cs]: >Assert [](start = 26, length = 6)

This check is fine I want to point we can do that too by calling CharUnicodeInfo.GetUnicodeCategory and check if it is upper case category. I think your check is better here as this code is in debug build only and no perf concern with it.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexBoyerMoore.cs
Comment (tarekgh) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexBoyerMoore.cs]: >textinfo.ToLower(text[index + i]) [](start = 38, length = 33)

wouldn't be faster if we convert the whole text (with the length of pattern) to lowercase as one shot? calling ToLower with every character can be very expensive if the character is not ASCII. 

The other idea is to call CharUnicodeInfo.GetUnicodeCategory on the char before calling tolower so we avoid teh overhead of characters that really don't need to be lower cased.

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexBoyerMoore.cs]: > wouldn't be faster if we convert the whole text (with the length of pattern) to lowercase as one shot?

How do we do that in a way that produces the same result and doesn't allocate?

> The other idea is to call CharUnicodeInfo.GetUnicodeCategory on the char before calling tolower so we avoid the overhead of characters that really don't need to be lower cased.

That will presumably add overhead for characters that do need to be lower-cased, correct?  Or else why wouldn't we just do that as part of the ToLower implementation?

Comment (tarekgh) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexBoyerMoore.cs]: > How do we do that in a way that produces the same result and doesn't allocate?

I expect the pattern length most of the time is small length (less than 200 characters or so). we can stackalloc and Span to do that I guess. 

> That will presumably add overhead for characters that do need to be lower-cased, correct? Or else why wouldn't we just do that as part of the ToLower implementation?

It depends if all characters we are using are uppercased. I am thinking the number of characters in Unicode that need casing compared to the whole set of Unicode characters is a really small percentage (and we'll be interested in the uppercased letters only too). 
Also CharUnicodeInfo.GetUnicodeCategory would cost a very small percentage compared to the time in ToLower (in non-ASCII cases).

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexBoyerMoore.cs]: > I expect the pattern length most of the time is small length (less than 200 characters or so). we can stackalloc and Span to do that I guess

And with what API?

Comment (tarekgh) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexBoyerMoore.cs]: > And with what API? i.e. what API does culture-aware lower-casing on a whole span/string with behavior equivalent to doing it character-by-character?

Currently, the casing (including the culture-aware one) always done in one-to-one matching the number of characters in the string. It doesn't collapse or expand any character. We can add simple API TextInfo.ToLower which works with the Span and I believe these APIs will be very useful in many other cases. 
By the way, you don't have to block the PR for that. I am just pointing we can optimize more in this case.

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexBoyerMoore.cs]: > We can add simple API TextInfo.ToLower which works with the Span and I believe these APIs will be very useful in many other cases.

If you think they'd be useful in other circumstances, I have no qualms with such methods being added.  Please open an issue for that, and if they're added, part of that work item should be updating this and any other possible call sites.

I was confirming that no such methods exist today, for span or string or any other input, correct?

Comment (tarekgh) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexBoyerMoore.cs]: > I was confirming that no such methods exist today, for span or string or any other input, correct?

I am not aware of any although I recall this was requested before but I didn't find any issue for that.

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexBoyerMoore.cs]: Great, thanks.

**Review by tarekgh — APPROVED** [review]
I'll leave it to you if you want to use CharUnicodeInfo.GetUnicodeCategory.

**Comment by stephentoub** [comment]
> I'll leave it to you if you want to use CharUnicodeInfo.GetUnicodeCategory.

That's a fine thing to experiment with for perf subsequently.  The change as it currently exists is a pure revert back to the release code we had before.  I want to fix the bug on its own, and then we could consider / implement / measure / tweak for perf.  There are many other places in the implementation that use ToLower; we shouldn't do this one in isolation.

**Comment by stephentoub** [comment]
Thanks for the review.

**Comment by danmosemsft** [comment]
@stephentoub should this be ported to 3.1?

**Comment by stephentoub** [comment]
> should this be ported to 3.1?

I'd be a little concerned about the breaking potential of this in an LTS release.  I'd leave it up to the area owners.

**Comment by tarekgh** [comment]
I prefer waiting not servicing this till we get a real request for it.

**Comment by danmosemsft** [comment]
I agree, just wanted to make sure it was noted.

**Pull request merged**
